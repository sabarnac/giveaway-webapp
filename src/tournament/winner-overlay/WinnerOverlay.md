Usage:

```tsx
import Participant from "./store/round/match/participant/Participant";

const winner: Participant = new Participant("FOO BAR");
<WinnerOverlay winner={winner} />;
```
