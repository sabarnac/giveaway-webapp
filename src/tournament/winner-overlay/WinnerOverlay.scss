.winner-overlay-wrapper {
    z-index: 1;
    background: rgb(0, 0, 0);
    position: fixed;
    top: 0;
    left: 0;
    height: 100vh;
    width: 100vw;
    overflow: hidden;
    display: flex;
    align-items: center;
    justify-content: center;
    opacity: 0.000001;
    transition: opacity 500ms ease-in-out;
    will-change: transition;
    &--exiting,
    &--exited {
        opacity: 0.000001;
    }
    &--entered,
    &--entering {
        opacity: 0.99999;
    }
    .winner-overlay {
        border: 1px dashed #000000;
        position: relative;
        background: #ffffff;
        &__winner {
            padding: 4rem 4rem 2rem 4rem;
            display: flex;
            flex-direction: column;
            align-items: center;
            &>.participant-card {
                margin-bottom: 3rem;
            }
        }
    }
}