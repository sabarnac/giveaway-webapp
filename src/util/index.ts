export let isDevEnvironment = (): boolean =>
  process.env.NODE_ENV === "development";

export let createAvatarImage = (identifier: string): string => {
  return `https://api.adorable.io/avatars/480/${identifier}`;
};

export let isInRange = (num: number, start: number, end: number): boolean =>
  num >= start && num <= end;
