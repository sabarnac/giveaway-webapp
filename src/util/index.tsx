import React, { ComponentType } from "react";
import { Dispatch, useState } from "react";
import Config from "../store/config/Config";
import { Redirect } from "react-router";
import { Observer } from "mobx-react";

/**
 * Returns if it is currently a dev environment.
 * @return {boolean} Whether it is currently a dev environment.
 */
export const isDevEnvironment = (): boolean =>
  process.env.NODE_ENV === "development";

/**
 * Returns a random avatar image URL.
 * @param {string} identifier The identifier to generate the avatar URL against.
 * @return {string} The avatar URL.
 */
export const createAvatarImage = (identifier: string): string => {
  return `https://api.adorable.io/avatars/480/${identifier}`;
};

/**
 * Checks if the number is within the range.
 * @param {number} num The number to check.
 * @param {number} start The start of the range.
 * @param {number} end The end of the range.
 * @return {boolean} Whether the number is within the range.
 */
export const isInRange = (num: number, start: number, end: number): boolean =>
  num >= start && num <= end;

/**
 * Type for custom animation state hook.
 */
export type AnimationStateHook = [
  number,
  () => void,
  (delay: number) => void,
  (state: number) => void
];

/**
 * Type for show overlay state hook.
 */
export type ShowOverlayHook = [boolean, (newValue: boolean) => void];

/**
 * Type for void function.
 */
export type VoidFunction = () => void;

/**
 * Runs the given action at least after the given delay.
 * @param {Function} action The action to run.
 * @param {number} delay The minimum delay to wait for.
 */
export const runOnDelay = (action: Function, delay: number): number =>
  setTimeout(action, delay);

/**
 * Create an animation state hook.
 * @return {AnimationStateHook} The current animation state, a state update method, and a delayed state update method.
 */
export const useAnimationState = (start: number = 0): AnimationStateHook => {
  const [currentState, setState]: [number, Dispatch<number>] = useState(start);

  const updateState = (): void => setState(currentState + 1);
  const updateStateDelay = (delay: number): VoidFunction => {
    const delayId: number = runOnDelay(updateState, delay);
    return () => clearTimeout(delayId);
  };

  return [currentState, updateState, updateStateDelay, setState];
};

/**
 * Normalize the given animation time by the animation speed multiplier.
 * @param {number} time The animation time to normalize.
 * @return {number} The normalized animation speed
 */
export const getNormalizedSpeed = (time: number): number =>
  time / Config.getInstance().speed;

/**
 * Runs the given action if the predicate is true.
 * @param {boolean} predicate The predicate to check.
 * @param {Function} action The action to run.
 */
export const runOnPredicate = (
  predicate: boolean,
  action: Function,
): VoidFunction | (() => VoidFunction) => (): void | VoidFunction =>
  predicate ? action() : undefined;

/**
 * Returns a redirect component for a given round and match.
 * @param {string} roundId The ID of the round to redirect to.
 * @param {string} matchId The ID of the match to redirect to.
 * @return {JSX.Element} The redirect element.
 */
export const getMatchRedirect = (roundId?: string, matchId?: string) => (
  <Redirect to={`/round/${roundId}/match/${matchId}`} />
);

/**
 * Returns a HOC with the given component wrapped by a MobX Observer component.
 * @param {ComponentType<T>} Component The component that should be under an observer.
 * @returns {(props: T) => JSX.Element} A HOC that wraps the given component under an observer.
 */
export const createObserver = <T extends {}>(
  Component: ComponentType<T>,
): ((props: T) => JSX.Element) => (props: T): JSX.Element => (
  <Observer>{() => <Component {...props} />}</Observer>
);
