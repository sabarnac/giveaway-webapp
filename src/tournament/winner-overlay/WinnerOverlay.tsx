import React, { useEffect } from "react";
import { Observer } from "mobx-react";
import "./WinnerOverlay.scss";
import Participant from "../../store/round/match/participant/Participant";
import classNames from "classnames";
import Config from "../../store/config/Config";
import { CSSTransition } from "react-transition-group";
import {
  useAnimationState,
  AnimationStateHookResult,
  getNormalizedSpeed,
} from "../../util";
import WinnerOverlayView from "./_partial/WinnerOverlayView";

/**
 * Properties of the winner overlay React component.
 */
interface WinnerOverlayProps {
  /** @ignore The application config. */
  config?: Config;
  /** The tournament winner. */
  winner: Participant;
}

/**
 * React component for the winner overlay.
 */
export default (props: WinnerOverlayProps): JSX.Element => {
  const [
    currentState,
    updateState,
  ]: AnimationStateHookResult = useAnimationState();
  const className: string = "winner-overlay-wrapper";

  useEffect(updateState);

  return (
    <Observer>
      {() => (
        <CSSTransition
          in={currentState > 0}
          timeout={getNormalizedSpeed(500)}
          classNames={{
            enter: "",
            enterActive: `${className}--entering`,
            enterDone: `${className}--entered`,
            exit: "",
            exitActive: `${className}--exiting`,
            exitDone: `${className}--exited`,
          }}
          mountOnEnter={true}
          unmountOnExit={true}
        >
          <div
            className={classNames(`${className}-wrapper`)}
            style={{
              transition: `opacity ${getNormalizedSpeed(500)}ms ease-in-out`,
            }}
          >
            <WinnerOverlayView winner={props.winner} />
          </div>
        </CSSTransition>
      )}
    </Observer>
  );
};
