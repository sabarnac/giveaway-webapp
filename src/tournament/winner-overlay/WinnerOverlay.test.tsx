import React from "react";
import WinnerOverlay from "./WinnerOverlay";
import { create, ReactTestRenderer } from "react-test-renderer";
import { createDummyParticipant } from "../../util/test";

describe("WinnerOverlay Component.", () => {
  it("Matches snapshot.", () => {
    const component: ReactTestRenderer = create(
      <WinnerOverlay show={true} winner={createDummyParticipant(1)} />,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
