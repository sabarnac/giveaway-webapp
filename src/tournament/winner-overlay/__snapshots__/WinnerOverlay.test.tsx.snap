// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WinnerOverlay Component. Matches snapshot. 1`] = `
<div
  classNames={
    Object {
      "enter": "",
      "enterActive": "winner-overlay-wrapper--entering",
      "enterDone": "winner-overlay-wrapper--entered",
      "exit": "",
      "exitActive": "winner-overlay-wrapper--exiting",
      "exitDone": "winner-overlay-wrapper--exited",
    }
  }
  in={false}
  mountOnEnter={true}
  timeout={500}
  unmountOnExit={true}
>
  <div
    className="winner-overlay-wrapper"
    style={
      Object {
        "transition": "opacity 500ms ease-in-out",
      }
    }
  >
    <div
      className="winner-overlay"
    >
      <div
        className="winner-overlay__winner"
      >
        <div
          className="participant participant-card"
        >
          <div
            className="participant-card__avatar"
          >
            <img
              alt="fozbaz-1"
              src="https://api.adorable.io/avatars/480/fozbaz-1"
            />
          </div>
          <div
            className="participant-card__name"
          >
            <strong>
              Fozbaz 1
            </strong>
          </div>
        </div>
        <h3>
          Won The Tournament!
        </h3>
      </div>
    </div>
  </div>
</div>
`;
